README.md
# HttpProxy

![Elixir CI](https://github.com/KazuCocoa/http_proxy/workflows/Elixir%20CI/badge.svg?branch=main)
[![](https://img.shields.io/hexpm/v/http_proxy.svg?style=flat)](https://hex.pm/packages/http_proxy)

[![codecov](https://codecov.io/gh/KazuCocoa/http_proxy/branch/main/graph/badge.svg)](https://codecov.io/gh/KazuCocoa/http_proxy)

Simple multi HTTP Proxy using Plug. And support record/play requests.

# MY GOAL
- Record/Play proxied requests
    - http_proxy support multi port and multi urls on one execution command `mix proxy`.
- Support VCR

# architecture

```
           http_proxy
Client  (server  client)  proxied_server
  |            |            |
  | 1.request  |            |
  |  ------>   | 2.request  |
  |            |  ------>   |
  |            |            |
  |            | 3.response |
  | 4.response |  <------   |
  |  <------   |            |
  |            |            |
```

1. The client sends a request to http_proxy, then the http_proxy works as a proxy server.
2. When the http_proxy receives the request from the client, then the http_proxy sends the request to a proxied server, e.g. http://google.com, as a client.
3. The http_proxy receives responses from the proxied_server, then the http_proxy sets the response into its response to the client.
4. The Client receives responses from the http_proxy.

# Quick use as http proxy
## requirement
- Elixir over 1.7

## set application and deps

- `mix.exs`
    - `:logger` is option.
    - `:http_proxy` is not need if you run http_proxy with `HttpProxy.start/0` or `HttpProxy.stop/0` manually.

```elixir
def application do
  [applications: [:logger, :http_proxy]]
end

...

defp deps do
  [
    {:http_proxy, "~> 1.4.0"}
  ]
end
```

## set configuration

- `config/config.exs`

```
use Mix.Config

config :http_proxy,
  proxies: [
             %{port: 8080,
               to:   "http://google.com"},
             %{port: 8081,
               to:   "http://yahoo.com"}
            ]
```

- To manage logger, you should define logger settings like the following.

```
config :logger, :console,
  level: :info
```

## solve deps and run a server

```
$ mix deps.get
$ mix clean
$ mix run --no-halt # start proxy server
```

If you would like to start production mode, you should run with `MIX_ENV=prod` like the following command.

```
$ MIX_ENV=prod mix run --no-halt
```

## launch browser

Launch browser and open `http://localhost:8080` or `http://localhost:8081`.
Then, `http://localhost:8080` redirect to `http://google.com` and `http://localhost:8081` do to `http://yahoo.com`.

# Development
- Copy `pre-commit` hook
    - `cp hooks/pre-commit ./git/hooks/pre-commit`

# Configuration
## Customize proxy port

- You can customize a proxy port. For example, if you change a waiting port from `8080` to `4000`, then you can access to `http://google.com` via `http://localhost:4000`.

```
use Mix.Config

config :http_proxy,
  proxies: [
             %{port: 4000,
               to:   "http://google.com"},
             %{port: 8081,
               to:   "http://yahoo.com"}
            ]
```

## Add proxy

- You can add a waiting ports in configuration file. For example, the following setting allow you to access to `http://apple.com` via `http://localhost:8082` in addition.

```
use Mix.Config

config :http_proxy,
  proxies: [
             %{port: 8080,
               to:   "http://google.com"},
             %{port: 8081,
               to:   "http://yahoo.com"},
             %{port: 8082,
               to:   "http://apple.com"}
            ]
```

## Play and Record mode

- When `:record` and `:play` are `false`, then the http_proxy works just multi port proxy.
- When `:record` is `true`, then the http_proxy works to record request which is proxied.
- When `:play` is `true`, then the http_proxy works to play request between this the http_proxy and clients.
    - You should set JSON files under `mappings` in `play_path`.
    - `config.proxies.to` must be available URL to succeed generating http client.
        - https://github.com/KazuCocoa/http_proxy/blob/main/lib/http_proxy/handle.ex#L49

```elixir
use Mix.Config

config :http_proxy,
  proxies: [                   # MUST
             %{port: 8080,     # proxy all request even play or record
               to:   "http://google.com"},
             %{port: 8081,
               to:   "http://yahoo.com"}
            ]
  timeout: 20_000,             # Option, ms to wait http request.
  record: false,               # Option, true: record requests. false: don't record.
  play: false,                 # Option, true: play stored requests. false: don't play.
  export_path: "test/example", # Option, path to export recorded files.
  play_path: "test/data"       # Option, path to read json files as response to.
```

## Example
### Record request as the following

```json
{
  "request": {
    "headers": [],
    "method": "GET",
    "options": {
      "aspect": "query_params"
    },
    "remote": "127.0.0.1",
    "request_body": "",
    "url": "http://localhost:8080/hoge/inu?email=neko&pass=123"
  },
  "response": {
    "body_file": "path/to/body_file.json",
    "cookies": {},
    "headers": {
      "Cache-Control": "public, max-age=2592000",
      "Content-Length": "251",
      "Content-Type": "text/html; charset=UTF-8",
      "Date": "Sat, 21 Nov 2015 00:37:38 GMT",
      "Expires": "Mon, 21 Dec 2015 00:37:38 GMT",
      "Location": "http://www.google.com/hoge/inu?email=neko&pass=123",
      "Server": "sffe",
      "X-Content-Type-Options": "nosniff",
      "X-XSS-Protection": "1; mode=block"
    },
    "status_code": 301
  }
}
```
Response body will save in "path/to/body_file.json".

### Play request with the following JSON data

- Example is https://github.com/KazuCocoa/http_proxy/tree/master/test/data/mappings
- You can set `path` or `path_pattern` as attribute under `request`.
    - If `path`, the http_proxy check requests are matched completely.
    - If `path_pattern`, the http_proxy check requests are matched with Regex.
- You can set `body` or `body_file` as attribute under `response`.
    - If `body`, the http_proxy send the body string.
    - If `body_file`, the http_proxy send the body_file binary as response.

#### `path` and `body` case

```json
{
  "request": {
    "path": "/request/path",
    "port": 8080,
    "method": "GET"
  },
  "response": {
    "body": "<html>hello world</html>",
    "cookies": {},
    "headers": {
      "Content-Type": "text/html; charset=UTF-8",
      "Server": "GFE/2.0"
    },
    "status_code": 200
  }
}
```

#### `path_pattern` and `body_file` case

- Pattern match with `Regex.match?(Regex.compile!("\A/request.*neko\z"), request_path)`
- `File.read/2` via `file/to/path.json` and respond the binary

```json
{
  "request": {
    "path_pattern": "\A/request.*neko\z",
    "port": 8080,
    "method": "GET"
  },
  "response": {
    "body_file": "file/to/path.json",
    "cookies": {},
    "headers": {
      "Content-Type": "text/html; charset=UTF-8",
      "Server": "GFE/2.0"
    },
    "status_code": 200
  }
}
```

## dependencies

```
$ mix xref graph
lib/http_proxy.ex
└── lib/http_proxy/supervisor.ex
    ├── lib/http_proxy/agent.ex
    │   ├── lib/http_proxy/play/data.ex
    │   │   ├── lib/http_proxy/agent.ex
    │   │   └── lib/http_proxy/play/response.ex
    │   │       ├── lib/http_proxy/play/data.ex
    │   │       └── lib/http_proxy/utils/file.ex
    │   └── lib/http_proxy/play/paths.ex
    │       ├── lib/http_proxy/agent.ex
    │       └── lib/http_proxy/play/response.ex
    └── lib/http_proxy/handle.ex
        ├── lib/http_proxy/play/body.ex
        ├── lib/http_proxy/play/data.ex
        ├── lib/http_proxy/play/paths.ex
        ├── lib/http_proxy/play/response.ex
        └── lib/http_proxy/record/response.ex
            ├── lib/http_proxy/format.ex
            │   └── lib/http_proxy/data.ex (compile)
            └── lib/http_proxy/utils/file.ex
```

# TODO
- [x] record request
- [x] play request
- [x] refactor
- [x] support Regex request path.
- [x] start/stop http_proxy manually
- [ ] ~~use vcr~~
    - integrate https://github.com/parroty/exvcr

# styleguide

http://elixir.community/styleguide

# LICENSE
MIT. Please read LICENSE.


CHANGELOG.md
# Changelogs
## 1.6.0: Nov 30, 2022
- Drop Elixir 1.11 and below

## 1.5.0: Apr 21, 2020
- Drop Elixir 1.7

## 1.4.0: Oct 21, 2018
- Update to Plug 1.7
    - `{:plug_cowboy, "~> 2.0"}`

## 1.3.1: May 20, 2018
- Update to Cowboy2 and Plug 1.5

## 1.3.0: Feb 17, 2018
- Support Elixir 1.6 and drop 1.5 and 1.4
    - Since supervisor stuff

## 1.2.3: Aug 28, 2017
- Update some dependencies

## 1.2.2: Jul 23, 2017
- Set extra_applications and runtime: false

## 1.2.1: Jul 23, 2017
- Update some dependencies
    - Set a limitation for Plug ~> 1.3.0

## 1.2.0: Jun 26, 2017
- Suppoer Elixir1.4.0+ and fix some warnings [here](https://github.com/KazuCocoa/http_proxy/pull/41)

## 1.1.5: Apr 6, 2017
- Update some dependencies

## 1.1.4: Mar 9, 2017
- Update some dependencies
- Remove ExVCR since this library isn't used in this project

## 1.1.3: Dec 1, 2016
- Update Plug to 1.3.0

## 1.1.2: Nov 21, 2016
- remove `export_path` and `play_path` from `%HttpProxy.Utils.File{}`
  - read them directly from `config/config.exs`

## 1.1.1: Nov 19, 2016
- update some dependencies

## 1.1.0: Aug 31, 2016
- Require Elixir ~> 1.3 due to update Plug from 1.1 to 1.2

## 1.0.3: Jul 5, 2016
- fix https://github.com/KazuCocoa/http_proxy/issues/25

## 1.0.2: Apr 14, 2016
- update some dependencies

## 1.0.1: Feb 1, 2016
- update some dependencies

## 1.0.0: Jan 16, 2016
- Lock over Elixir1.2
- Support store play responses in Agent.

## 0.6.0: Jan 7, 2016
- support setting response body from file.
- read README and `body_file`

## 0.5.3: Dec 31, 2015
- Fix getting headers in play mode with Elixir 1.2.0 [#15](https://github.com/KazuCocoa/http_proxy/issues/15)

## 0.5.1 and 0.5.2: Dec 30, 2015
- Add `HttpProxy.stop` and `HttpProxy.start` to stop/start the proxy manually.

## 0.5.0: Dec 29, 2015
- Support regex path matching in play mode.

You can specify request path with Regex like the following `path_pattern`.

```
{
  "request": {
    "path_pattern": "\A/request.*neko\z",
    "port": 8080,
    "method": "GET"
  },
  ...
}
```

And some refactors.

## 0.4.1: Dec 25, 2015
record timeout into jsons when request timeout [#3](https://github.com/KazuCocoa/http_proxy/issues/3)

## 0.4.0: Dec 23, 2015
Export file into `mappings` / `__files` with recording mode

e.g.

1. When record a request and a response
2. Then two files are saved into following files
    - `test/example/8080/mappings/request_file.json`
    - `test/example/8080/__files/request_file.json`

## 0.3.4: Dec 23, 2015
- update timeout settings

## 0.3.3: Dec 21, 2015
- Support timeout option
- Record every request/response body

## 0.3.2: Dec 18, 2015
- fix define method instead some initial value [#1](https://github.com/KazuCocoa/http_proxy/issues/1)
- fix don't sent request when proxy mode

## 0.3.0: Dec 8, 2015
- Support record/play request for each proxy.

## 0.2.1: Nov 21, 2015
- arrange format like VRC one.

## 0.2.0: Nov 13, 2015
- support record request and export into JSON

## 0.1.0: Nov 8, 2015
- initial release


lib/http_proxy.ex
defmodule HttpProxy do
  @moduledoc """
  HttpProxy is a simple http proxy.

  If you access to particular URL like `http://localhost:8080`, then the http_proxy forward the request to other URL
  based on configuration.

  HttpProxy supports two features.

  1. HttpProxy supports a multi-port proxy.
  2. HttpProxy supports play/record proxied requests.

  Multiport proxy means that the proxy receives request with particular port and the proxy send request to other address.
  And you can set the feature against several mulatiple port.

  # Multiport proxy

  For example, you set configuratio like the followings in your project and do `mix proxy`. Then the proxy send request to
  "http://google.com" if anyone snds to "http://localhost:4000". And the proxy send request to "http://yahoo.com" if anyone
  send request to "http://localhost:4001".

  ## example
  1. Set configuration as the following in `config/config.exs`.

          use Mix.Config

          config :http_proxy,
          proxies: [
                     %{port: 4000,
                       to:   "http://google.com"},
                     %{port: 4001,
                       to:   "http://yahoo.com"}
                    ]
          record: false,
          play: true,
          export_path: "test/example",
          play_path: "test/data"

  2. Access to `http://localhost:4000` via Web Browser.
      - or access to  `http://localhost:4001`.
  3. The http_proxy forward to **http://google.com** .
      - or forward to  **http://yahoo.com**

  # Play/Record proxied request.

  If you set `record: true` in the configuration, the proxy export request into local file as JSON.
  You can export requests in particular path which is set as `export_path: "test/example"`. Default is "default".

  If you set `play: true` in the configuration, the proxy read mapping files and reply them when anyone accesses
  to the proxy via particular ports.

  Please read `test/data/mappings/*.json` if you would like to know the format of playing the reqponse.
  """

  use Application
  alias HttpProxy.Supervisor, as: HttpProxySup

  @spec start(:normal, []) :: {:ok, pid}
  def start(_type, _args) do
    HttpProxySup.start_link()
  end

  @doc """
  Stopt http_proxy.
  """
  @spec stop :: :ok | {:error, term}
  def stop do
    Application.stop(:ranch)
    Application.stop(:cowlib)
    Application.stop(:cowboy)
    Application.stop(:idna)
    Application.stop(:mimerl)
    Application.stop(:certifi)
    Application.stop(:hackney)
    Application.stop(:plug)
    Application.stop(:http_proxy)
  end

  @doc """
  Start http_proxy.
  If the proxy is already running, return `{:error, {:already_started, :http_proxy}}`
  """
  @spec start() :: :ok | {:error, term}
  def start do
    Application.start(:ranch)
    Application.start(:cowlib)
    Application.start(:cowboy)
    Application.start(:idna)
    Application.start(:mimerl)
    Application.start(:certifi)
    Application.start(:hackney)
    Application.start(:plug)
    Application.start(:http_proxy)
  end
end
